cratedb_toolkit/wtf/library.py

```diff
@@ -0,0 +1,587 @@
+            heap['used'] / heap['max']::DOUBLE AS heap_used
+        FROM sys.nodes
+        ORDER BY node_name;
+        """
+
+        tcp_connections = """
+        SELECT
+            name AS node_name,
+            connections
+        FROM sys.nodes
+        ORDER BY node_name;
+        """
+
+        # TODO: Why "14"? Is it about only getting information about the `write` thread pool?
+        thread_pools = """
+        SELECT
+            name AS node_name,
+            thread_pools[14]['queue'],
+            thread_pools[14]['active'],
+            thread_pools[14]['threads']
+        FROM sys.nodes
+        ORDER BY node_name;
+        """
+
+    class Settings:
+        """
+        Reflect cluster settings.
+        """
+
+        info = """
+        SELECT
+            name,
+            master_node,
+            settings['cluster']['routing']['allocation']['cluster_concurrent_rebalance']
+                AS cluster_concurrent_rebalance,
+            settings['indices']['recovery']['max_bytes_per_sec'] AS max_bytes_per_sec
+        FROM sys.cluster
+        LIMIT 1;
+        """
+
+    class Shards:
+        """
+        Information about shard / node / table / partition allocation and rebalancing.
+        """
+
+        # https://cratedb.com/docs/crate/reference/en/latest/admin/system-information.html#example
+        # TODO: Needs templating.
+        for_table = """
+        SELECT
+            schema_name,
+            table_name,
+            id,
+            partition_ident,
+            num_docs,
+            primary,
+            relocating_node,
+            routing_state,
+            state,
+            orphan_partition
+        FROM sys.shards
+        WHERE schema_name = '{schema_name}' AND table_name = '{table_name}';
+        """
+
+        # Identify the location of the shards for each partition.
+        # TODO: Needs templating.
+        location_for_partition = """
+        SELECT   table_partitions.table_schema,
+                 table_partitions.table_name,
+                 table_partitions.values[{partition_column}]::TIMESTAMP,
+                 shards.primary,
+                 shards.node['name']
+        FROM sys.shards
+        JOIN information_schema.table_partitions ON shards.partition_ident=table_partitions.partition_ident
+        WHERE table_partitions.table_name = {table_name}
+        ORDER BY 1,2,3,4,5;
+        """
+
+        allocation = InfoElement(
+            name="shard_allocation",
+            sql="""
+                SELECT
+                    IF(s.primary = TRUE, 'primary', 'replica') AS shard_type,
+                    COALESCE(shards, 0) AS shards
+                FROM
+                    UNNEST([true, false]) s(primary)
+                LEFT JOIN (
+                  SELECT primary, COUNT(*) AS shards
+                  FROM sys.allocations
+                  WHERE current_state != 'STARTED'
+                  GROUP BY 1
+                ) a ON s.primary = a.primary;
```

Comment:
This looks unreasonable complicated to just translate the primary boolean into a string.
```suggestion
                SELECT
                    IF(primary = TRUE, 'primary', 'replica') AS shard_type,
                    COUNT(*) AS shards
                  FROM sys.allocations
                  WHERE current_state != 'STARTED'
                  GROUP BY 1
```

Reply:
See GH-169.

Reply:
As I wrote a suggestion already, wouldn't it be feasible to change this query right in this PR? I don't see much value in postponing this, maybe my suggestions is lost then..

Reply:
I've applied your suggestion 1:1, with ef5438b118. Thanks.